### DISCUSSION TITLE
Implementing the modified Cam Clay model in Mfront

### DISCUSSION THREAD

**Discussion by kuateric** [thelfer/tfel#76] [General]
Hi @thelfer, 
I found the equations of the modified cam clay model in my book ready for implementation 
here the plastic multiplier: 
![image](https://user-images.githubusercontent.com/22998049/149086853-bb1d2c84-ff8f-41ba-a7d9-6c303b04cd16.png)

and here the elasto-plastic tangent operator: 
![image](https://user-images.githubusercontent.com/22998049/149086940-3e9e4b13-b8e2-4ebe-bfe0-f15d3f55c6de.png)

I want to give my shot and try to implement it and progressing in the learning curve
First question: does it make any difference if the tangent operator is not symmetric ? 
is there a `@ProvidesAssymmetricTangentOperator`

**Comment by thelfer** [comment]
Hi @kuateric,

You may also refer to <https://thelfer.github.io/MFrontGallery/web/SemiImplicitModifiedCamClay_OpenGeoSys2020.html> based on the Implicit DSL.

However, as you seem to have closed-form formulae, using the `Default` DSL seems much better. Quite surprised that Christian and Thomas did not follow this path. Maybe I missed something :)

Concerning your question, the best is to ask `mfront` for help:

```bash
$ mfront --help-keywords-list=Default |grep Operator
- @IsTangentOperatorSymmetric                (undocumented)
- @PredictionOperator                        (undocumented)
- @ProvidesSymmetricTangentOperator          (undocumented)
- @ProvidesTangentOperator                   (undocumented)
- @RequireStiffnessOperator                  (documented)
- @TangentOperator                           (documented)
```

which shows that the keyword you are looking for is `@ProvidesTangentOperator`.

Best,

Thomas

**Comment by kuateric** [comment]
ok thanks! 
I actually comparing both implementations. I have to stay that I don't understand the implementation in OGS ;), I still need some learning.  What is missing in the book implementation is the dependency of the volumetric stress component in function of the porosity. Also, this implementation seems to be more general than the David's implementation (more parameters). 
So I am trying to make a synthesis of the three models ;)
